### DISCUSSION TITLE
gotmpl in helmfile YAML files

### DISCUSSION THREAD

**Discussion by vquie** [oxsecurity/megalinter#2574] [Q&A]
Hi,

we are currently in the middle of lifting mega linter into production state to clean our code. Also we are now facing an issue where I think mega-linter treats gotmpl in yaml files as json.
`{{ … }}` would become `{ { … } }`. We have disabled linting on these files with the regex exclude filter to avoid breaking changes.
What is the best way, if any, to solve this?

We are using ci_light flavor with hosted GitLab on our own runners.

Cheers

**Comment by Kurt-von-Laven** [comment]
This question sounds like it might be best addressed to [helm](https://github.com/helm/helm).

**Reply by vquie** [reply]
It’s actually helmfile.

https://github.com/roboll/helmfile/blob/master/docs/writing-helmfile.md

Do you mean I should ask over there?

**Reply by Kurt-von-Laven** [reply]
Yeah, I think you will have better odds. I don't see a connection to MegaLinter's behavior.

**Reply by vquie** [reply]
I thought we could treat the file accordingly to the extension and not to the content.
Or maybe keep json linters out if it a yaml, or maybe dig a little deeper if it that case after all and not something else is going on.

**Reply by Kurt-von-Laven** [reply]
MegaLinter only uses extensions, not content, to decide which linters to invoke, so my assumption is that Helm is modifying your Helm files, not some other linter. To test this, you could try running with only Helm enabled or run Helm directly.

**Reply by nvuillam** [reply]
@Kurt-von-Laven MegaLinter can also use content , if property file_contains is present on the linter descriptor :)

**Comment by nvuillam** [comment]
@vquie the conditions for detecting helm file are the following: https://megalinter.io/latest/descriptors/kubernetes_helm/#how-are-identified-applicable-files

- Activated only if sub-directory ` is found. (directory name can be overridden withKUBERNETES_DIRECTORY`)
- Activated only if one of these files is found: Chart.yml, Chart.yaml
- File extensions: .yml, .yaml, .json
- Detected file content (regex): apiVersion:, kustomize\.config\.k8s\.io, tekton

What linter impacts which files in your repo ?
I'm not very familiar with gotmpl , please give more info so we can help :)

**Reply by vquie** [reply]
It is helmfile, not helm. I found out that prettier is causing this and it is known. The only workaround is to disable the [bracketSpacing](https://prettier.io/docs/en/options.html#bracket-spacing) rule. Then it will remove spaces that are set to improve readability though. The only solution for us was to disable prettier on basically every yaml file.

**Reply by nvuillam** [reply]
There is probably a single prettier rule to tweak or disable in a prettier config file
Anyway... I think it's not a big deal to not use it in yaml files, as IDE extensions can autoformat it, and at worse yamllint will catch issues :)

**Reply by Kurt-von-Laven** [reply]
You could alternatively use `YAML_PRETTIER_FILTER_REGEX_EXCLUDE` to disable Prettier only on certain YAML files.

**Reply by vquie** [reply]
Thats what I did actually.

```
YAML_PRETTIER_FILTER_REGEX_EXCLUDE: (helmfiles)
```

Now every yaml file in my helmfiles folder will be ignored.

Unfortunately `YAML_PRETTIER_FILTER_REGEX_EXCLUDE ` does not follow the same rules as `FILTER_REGEX_EXCLUDE`. I.e. `.*\/helmfiles\/.*\.ya?ml` would render to `.*\\/helmfiles\\/.*\\.ya?ml` which obviously doesn't match any folders. A similar regex in `FILTER_REGEX_EXCLUDE` works as expected. Thats why I now just look for the parent dir.

**Reply by nvuillam** [reply]
That's very strange, the code called to filter files is the same at root, descriptor and linter level :/

**Reply by nvuillam** [reply]
Are you sure this is prettier that formats the file ?
Could you try with LOG_LEVEL: DEBUG to have more output so we can investigate ?

**Reply by vquie** [reply]
I am sure. I ran it with `LOG_LEVEL: DEBUG` and found the very command  (`prettier --write $file` in an array). Was able to reproduce it with this command. And also..... this false changes do not happen again since I disabled it with the line I sent earlier. :-)

**Reply by Kurt-von-Laven** [reply]
This is really a stretch, but one possible explanation could be that some other instance of Prettier, such as `JSON_PRETTIER`, is running on the Helm files and isn't filtered. Hard to say much more here without seeing the logs though.

**Reply by vquie** [reply]
Please have a look at this example: https://github.com/vquie/mega-linter-prettier-test
You can easily reproduce it and debug logs are pushed to the repo.
